crates/shielded_token/src/vp.rs

```diff
@@ -128,8 +226,12 @@ where
     fn valid_nullifiers_reveal(
         ctx: &'ctx CTX,
         keys_changed: &BTreeSet<Key>,
-        transaction: &Transaction,
+        masp_source: &MaspSource,
     ) -> Result<()> {
+        let MaspSource::UserTx(transaction) = masp_source else {
```

Comment:
I believe here we are making an assumption on the correct behavior of the wasm transaction. As it stands right now this early exit is correct but in case of a buggy or malicious tx that reveals unnecessary nullifiers we would skip the check at line 274 and accept an invalid tx

Reply:
The masp vp was already assuming some collaboration from wasm txs, by means of the actions wasm protocol, which is used to inform the masp vp of shielded tx section references, and bundle signers. I definitely agree with you, though, just pointing out this kind of flaw isn't new. In any case, the chances of a nullifier collision due to this are pretty low, $1/(2^{32 \times 8} - 1)$. I'd say relying on wasm whitelisting for some degree of security is fine.